### ISSUE TITLE
Eval stable function of rangeFuncion during planning leads to error

### ISSUE THREAD

**Issue by kainwen** [greenplum-db/gpdb#14499]
### Greenplum version or build

* main branch (top commit d3da12fb12d1f97b031cbf90882c8e77bee902a2)
* 6X_STABLE (top commit b32e9d729883e280073355e6c5b23364e527c2a9)

### Step to reproduce the behavior

```sql
SELECT * FROM  json_populate_record(null::record, '{"x": 776}') AS (x int, y int);
ERROR:  could not determine row type for result of json_populate_record
HINT:  Provide a non-null record argument, or call the function in the FROM clause using a column definition list.
```
works normally in postgres12
```sql
psql (12.13)
Type "help" for help.

interma=# select version();
                                                     version
------------------------------------------------------------------------------------------------------------------
 PostgreSQL 12.13 on arm-apple-darwin21.6.0, compiled by Apple clang version 14.0.0 (clang-1400.0.29.102), 64-bit
(1 row)

interma=# SELECT * FROM  json_populate_record(null::record, '{"x": 776}') AS (x int, y int);
  x  | y
-----+---
 776 |
(1 row)
```

### RCA

Stable function is not evaluated until entering executor, this is normal behavior of Postgres. However, Greenplum has its own specific behavior, evaluate stable function will lead to much better OLAP performance, e.g. partition pruning can happened during planning stage. Thus Greenplum spend much work on making it correct.

`json_populate_record` is rangefunc here (stable), it should not be evaluated during planning stage.

**Comment by kainwen** [comment]
I have tried another idea (different from @interma 's ) to fix the issue (diff attached at the end): **not to evaluate stable range function**. It breaks `qp_functions_in_from.sql` case. Some of the cases of function scan join table now will not evaluate the function during planning, thus will be dispatched to QEs to execute and the function contains SQL to access distributed table. A typical failure case is below:

```
gpadmin=# explain (costs off) SELECT * FROM func1_read_int_sql_stb(func2_nosql_stb(5)), foo order by 1,2,3;
                                  QUERY PLAN
-------------------------------------------------------------------------------
 Gather Motion 2:1  (slice1; segments: 2)
   Merge Key: func1_read_int_sql_stb.func1_read_int_sql_stb, foo.a, foo.b
   ->  Sort
         Sort Key: func1_read_int_sql_stb.func1_read_int_sql_stb, foo.a, foo.b
         ->  Nested Loop
               ->  Broadcast Motion 1:2  (slice2)
                     ->  Function Scan on func1_read_int_sql_stb
               ->  Seq Scan on foo
 Optimizer: Postgres query optimizer
```

The above plan is generated based on the new method (not evaluate stable rangefunc). And the UDF `func1_read_int_sql_stb` contains PL to access distributed table. Now it will error out. The cases are there for a long time, such kind of usage is rare in practical env.

I am not sure which method should be used. I prefer to use @interma 's https://github.com/greenplum-db/gpdb/pull/14531 . There are little such functions in catalog we can use a disallow list. Until we hear more real complains then we design new ideas.


```diff
diff --git a/src/backend/optimizer/util/clauses.c b/src/backend/optimizer/util/clauses.c
index df3542624da..d1a9aeb5e9c 100644
--- a/src/backend/optimizer/util/clauses.c
+++ b/src/backend/optimizer/util/clauses.c
@@ -75,6 +75,7 @@ typedef struct
 	bool		recurse_queries; /* recurse into query structures */
 	bool		recurse_sublink_testexpr; /* recurse into sublink test expressions */
 	Size        max_size; /* max constant binary size in bytes, 0: no restrictions */
+	List       *rangefunc_oids; /* GPDB: only use when eval_stable_functions is true */
 } eval_const_expressions_context;

 typedef struct
@@ -168,7 +169,7 @@ static bool tlist_matches_coltypelist(List *tlist, List *coltypelist);
 /*
  * Greenplum specific functions
  */
-static bool should_eval_stable_functions(PlannerInfo *root);
+static bool should_eval_stable_functions(PlannerInfo *root, List **rangefunc_oids);

 /*****************************************************************************
  *		Aggregate-function clause manipulation
@@ -2358,7 +2359,7 @@ fold_constants(PlannerInfo *root, Query *q, ParamListInfo boundParams, Size max_

 	context.max_size = max_size;

-	context.eval_stable_functions = should_eval_stable_functions(root);
+	context.eval_stable_functions = should_eval_stable_functions(root, &(context.rangefunc_oids));

 	return (Query *) query_or_expression_tree_mutator
 						(
@@ -2493,7 +2494,7 @@ eval_const_expressions(PlannerInfo *root, Node *node)
 	context.recurse_queries = false; /* do not recurse into query structures */
 	context.recurse_sublink_testexpr = true;
 	context.max_size = 0;
-	context.eval_stable_functions = should_eval_stable_functions(root);
+	context.eval_stable_functions = should_eval_stable_functions(root, &(context.rangefunc_oids));

 	saved_oid_assignments = SaveOidAssignments();
 	result = eval_const_expressions_mutator(node, &context);
@@ -2534,6 +2535,7 @@ estimate_expression_value(PlannerInfo *root, Node *node)
 	context.recurse_sublink_testexpr = true;
 	context.max_size = 0;
 	context.eval_stable_functions = false;
+	context.rangefunc_oids = NIL; /* GPDB */

 	return eval_const_expressions_mutator(node, &context);
 }
@@ -4680,6 +4682,8 @@ evaluate_function(Oid funcid, Oid result_type, int32 result_typmod,
 		 /* okay */ ;
 	else if (context->eval_stable_functions && funcform->provolatile == PROVOLATILE_STABLE)
 	{
+		if (context->rangefunc_oids != NIL && list_member_oid(context->rangefunc_oids, funcid))
+			return NULL;
 		/* okay, but we cannot reuse this plan */
 		context->root->glob->oneoffPlan = true;
 	}
@@ -5796,8 +5800,9 @@ tlist_matches_coltypelist(List *tlist, List *coltypelist)
  * typical queries are long-running.
  */
 static bool
-should_eval_stable_functions(PlannerInfo *root)
+should_eval_stable_functions(PlannerInfo *root, List **rangefunc_oids)
 {
+	*rangefunc_oids = NIL;
 	/*
 	 * Without PlannerGlobal, we cannot mark the plan as a `oneoffPlan`
 	 */
@@ -5814,17 +5819,37 @@ should_eval_stable_functions(PlannerInfo *root)
 	if (root->query_level > 1) return true;

 	int rtable_num_total = list_length(root->parse->rtable);
+	List     *func_oids = NIL;
 	ListCell *lc;
 	foreach(lc, root->parse->rtable)
 	{
 		RangeTblEntry *rte = (RangeTblEntry *)lfirst(lc);

+		if (rte->rtekind == RTE_FUNCTION)
+		{
+			ListCell *cell;
+			foreach(cell, rte->functions)
+			{
+				RangeTblFunction *rtfunc = (RangeTblFunction *) lfirst(cell);
+				func_oids = lappend_oid(func_oids, ((FuncExpr *) (rtfunc->funcexpr))->funcid);
+			}
+		}
+
 		/* RTE_RESULT is a dummy RTE generated by empty from clause */
 		if (rte->rtekind == RTE_RESULT)
 			rtable_num_total --;
 	}

-	return rtable_num_total > 0;
+	if (rtable_num_total > 0)
+	{
+		*rangefunc_oids = func_oids;
+		return true;
+	}
+	else
+	{
+		*rangefunc_oids = NIL;
+		return false;
+	}
 }
```

**Comment by zxuejing** [comment]
this has been fixed by @interma.
merged in commit f5f6635da : Merge with PostgreSQL 12.12

**Comment by zxuejing** [comment]
evaluate_function
if (funcform->prorettype == ANYELEMENTOID && funcform->provolatile == PROVOLATILE_STABLE)
		if (in_stable_retany_funclist(funcform->oid))
			return NULL;
